devito/data/utils.py

```diff
@@ -331,3 +309,133 @@ def mpi_index_maps(loc_idx, shape, topology, coords, comm):
         local_si[index] = as_tuple(rnorm_index)
         it.iternext()
     return owners, send, global_si, local_si
+
+
+def flip_idx(idx, decomposition):
+    """
+    This function serves two purposes:
+    1) To Convert a global index with containing a slice with step < 0 to a 'mirrored'
+       index with all slice steps > 0.
+    2) Normalize indices with slices containing negative start/stops.
+
+    Parameters
+    ----------
+    idx: Tuple of slices/ints/tuples
+        Representation of the indices that require processing.
+    decomposition : tuple of Decomposition
+        The data decomposition, for each dimension.
+
+    Examples
+    --------
+    In the following examples, the domain consists of 12 indices, split over
+    four subdomains [0, 3]. We pick 2 as local subdomain.
+
+    >>> from devito.data import Decomposition, flip_idx
+    >>> d = Decomposition([[0, 1, 2], [3, 4], [5, 6, 7], [8, 9, 10, 11]], 2)
+    >>> d
+    Decomposition([0,2], [3,4], <<[5,7]>>, [8,11])
+
+    Example with negative stepped slices:
+
+    >>> idx = (slice(4, None, -1))
+    >>> fidx = flip_idx(idx, (d,))
+    >>> fidx
+    (slice(None, 5, 1),)
+
+    Example with negative start/stops:
+
+    >>> idx2 = (slice(-4, -1, 1))
+    >>> fidx2 = flip_idx(idx2, (d,))
+    >>> fidx2
+    (slice(8, 11, 1),)
+    """
+    processed = []
+    for i, j in zip(as_tuple(idx), decomposition):
+        if isinstance(i, slice) and i.step is not None and i.step < 0:
+            if i.start is None:
+                stop = None
+            elif i.start > 0:
+                stop = i.start + 1
+            else:
+                stop = i.start + j.glb_max + 2
+            if i.stop is None:
+                start = None
+            elif i.stop > 0:
+                start = i.stop + 1
+            else:
+                start = i.stop + j.glb_max + 2
+            processed.append(slice(start, stop, -i.step))
+        elif isinstance(i, slice):
+            if i.start is not None and i.start < 0:
+                start = i.start + j.glb_max + 1
+            else:
+                start = i.start
+            if i.stop is not None and i.stop < 0:
+                stop = i.stop + j.glb_max + 1
+            else:
+                stop = i.stop
+            processed.append(slice(start, stop, i.step))
+        else:
+            processed.append(i)
+    return as_tuple(processed)
+
+
+def distributed_data_size(shape, coords, topology):
+    """
+    Compute the cumulative shape of the distributed data (cshape).
+
+    coords: Array containing the local shape of data to each rank.
+
+    send: Array of rank coordinates corresponding to coords.
+
+    topology: Topology of the associated distributor.
+
+    Examples
+    --------
+    Given a set of distributed data such that:
```

Comment:
it's pretty much all there already...maybe we can turn this into a testable example, with `>>>` .. ?
if I'm asking too much, ignore me please

Reply:
Could do but there's just a lot of things to import and create so it gets a bit messy. I'd prefer to leave as is but can add if you'd really like.

Reply:
no it's OK, I get it, no worries. This is however heavily tested indirectly, so that's fine